dissect/cstruct/types/structure.py

```diff
@@ -397,11 +397,24 @@ class UnionMetaType(StructureMetaType):
     """Base metaclass for cstruct union type classes."""
 
     def __call__(cls, *args, **kwargs) -> Union:
-        obj = super().__call__(*args, **kwargs)
-        if kwargs:
-            # Calling with kwargs means we are initializing with values
-            # Proxify all values
+        obj: Union = super().__call__(*args, **kwargs)
+
+        # Calling with non-stream args or kwargs means we are initializing with values
+        if (args and not (len(args) == 1 and (_is_readable_type(args[0]) or _is_buffer_type(args[0])))) or kwargs:
+            # We don't support user initialization of dynamic unions yet
+            if cls.dynamic:
+                raise NotImplementedError("Initializing a dynamic union is not yet supported")
+
+            # User (partial) initialization, rebuild the union
+            # First user-provided field is the one used to rebuild the union
+            arg_fields = [field.name for _, field in zip(args, cls.__fields__)]
+            kwarg_fields = [name for name in kwargs if name in cls.lookup]
+            arg_fields.extend(kwarg_fields)
+            obj._rebuild(next(iter(arg_fields)))
```

Comment:
I think `arg_fields` can be an empty list at this point? 
If so, this might throw some exception, which is might be difficult for user to understand.

Reply:
I think we crash before we get here in the scenario where `arg_fields` could be empty, but changed it regardless.